### PR TITLE
[BugFix] Add relabel_nodes option for graph.subgraph function

### PR DESCRIPTION
**Pull request by caojy1998**
## Description

Add relabel_nodes option for graph.subgraph() function on CPU.
Also add a simple test case in test_subgraph.py
The related issue is: https://github.com/dmlc/dgl/issues/5880#issue-1759988303

## Checklist
Please feel free to remove inapplicable items for your PR.
- [ ] The PR title starts with [$CATEGORY] (such as [NN], [Model], [Doc], [Feature]])
- [ ] I've leverage the [tools](https://docs.google.com/document/d/1iHyj7zlmygKSk5gBPsqIqL5ASPzJSPREaNT_QdsiYA4/edit) to beautify the python and c++ code.
- [ ] The PR is complete and small, read the [Google eng practice (CL equals to PR)](https://google.github.io/eng-practices/review/developer/small-cls.html) to understand more about small PR. In DGL, we consider PRs with less than 200 lines of core code change are small (example, test and documentation could be exempted).
- [ ] All changes have test coverage
- [ ] Code is well-documented
- [ ] To the best of my knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change
- [ ] Related issue is referred in this PR
- [ ] If the PR is for a new model/paper, I've updated the example index [here](../examples/README.md).

### CODE REVIEW

File: examples/core/graphsage/node_classification.py [hunk 1]
```diff
@@ -99,7 +99,7 @@ def train(g, features, labels, masks, model):
     #
     # First, the whole graph is loaded and transformed. Then the training
     # process is performed on a model which is composed of 2 GraphSAGE-gcn
-    # layer. Finally, the performance of the model is evaluated on test set.
+    # layers. Finally, the performance of the model is evaluated on test set.
```
Comment (Rhett-Ying) [inline examples/core/graphsage/node_classification.py:102]: please do not include unrelated change such as typo fix. it's better to do in a separate PR.

**Review by Rhett-Ying — COMMENTED** [review]
How about adding the support for `relable_nodes=False` in the inner function first(in this PR), then change upper level calls(in following PRs).

File: tests/python/common/test_subgraph.py [hunk 2]
```diff
@@ -95,6 +95,49 @@ def test_subgraph():
     assert F.allclose(h, g.ndata["h"])
 
 
+def test_subgraph_relabel_nodes():
```
Comment (frozenbugs) [inline tests/python/common/test_subgraph.py:98]: test_subgraph_not_relabel_nodes?

Comment (frozenbugs) [reply tests/python/common/test_subgraph.py:98]: Maybe clarify the test_subgraph to test_subgraph_relabel_nodes, and change L62 to g.subgraph(nid, relabel_nodes=True)

**Comment by frozenbugs** [comment]
Can you link the related issue?

**Comment by caojy1998** [comment]
> Can you link the related issue?

OK, I have related it.

File: src/array/cpu/spmat_op_impl_csr.cc [hunk 3]
```diff
@@ -492,7 +492,55 @@ template CSRMatrix CSRSliceRows<kDGLCPU, int64_t>(CSRMatrix, NDArray);
 
 template <DGLDeviceType XPU, typename IdType>
 CSRMatrix CSRSliceMatrix(
-    CSRMatrix csr, runtime::NDArray rows, runtime::NDArray cols) {
+    CSRMatrix csr, runtime::NDArray rows, runtime::NDArray cols,
```
Comment (frozenbugs) [inline src/array/cpu/spmat_op_impl_csr.cc:495]: For the 2 cases, do we have to create 2 unrelated code block or can we actually share some part of the work?

If we have to create 2 big block, I'd do:
```
CSRMatrix CSRSliceMatrixWithNodesRelabeling() {...}
CSRMatrix CSRSliceMatrixWithOutNodesRelabeling() {...}

CSRMatrix CSRSliceMatrix(bool relabel_nodes) {
  return relabel_nodes ? CSRSliceMatrixWithNodesRelabeling() : CSRSliceMatrixWithOutNodesRelabeling();
}
```

File: tests/python/common/test_subgraph.py (hunk 2 above)
Comment (caojy1998) [reply tests/python/common/test_subgraph.py:98]: OK, I have modified these.

File: src/array/cpu/spmat_op_impl_csr.cc (hunk 3 above)
Comment (caojy1998) [reply src/array/cpu/spmat_op_impl_csr.cc:495]: OK, I try to make the change in this function simpler.

File: src/array/cpu/spmat_op_impl_csr.cc [hunk 4]
```diff
@@ -492,10 +492,17 @@ template CSRMatrix CSRSliceRows<kDGLCPU, int64_t>(CSRMatrix, NDArray);
...
+    bool relabel_nodes) {
+  int64_t new_nrows_temp = rows->shape[0];
+  int64_t new_ncols_temp = cols->shape[0];
+  if (relabel_nodes == false) {
```
Comment (frozenbugs) [inline src/array/cpu/spmat_op_impl_csr.cc:499]:
```suggestion
  if (!relabel_nodes) {
```

File: src/array/cpu/spmat_op_impl_csr.cc [hunk 5]
```diff
@@ -508,16 +515,21 @@ CSRMatrix CSRSliceMatrix(
...
     // NOTE: newi == i
     const IdType oldi = rows_data[i];
     CHECK(oldi >= 0 && oldi < csr.num_rows) << "Invalid row index: " << oldi;
     for (IdType p = indptr_data[oldi]; p < indptr_data[oldi + 1]; ++p) {
       const IdType oldj = indices_data[p];
       const IdType newj = hashmap.Map(oldj, kInvalidId);
       if (newj != kInvalidId) {
-        ++sub_indptr[i];
-        sub_indices.push_back(newj);
+        if (relabel_nodes == true) {
```
Comment (frozenbugs) [inline src/array/cpu/spmat_op_impl_csr.cc:526]:
```suggestion
        if (relabel_nodes) {
```

**Review by frozenbugs — APPROVED** [review]
LGTM, @Rhett-Ying please double check.

File: src/array/cpu/spmat_op_impl_csr.cc (hunk 4 above)
Comment (caojy1998) [reply src/array/cpu/spmat_op_impl_csr.cc]: OK, I have changed it.

File: src/array/cpu/spmat_op_impl_csr.cc (hunk 5 above)
Comment (caojy1998) [reply src/array/cpu/spmat_op_impl_csr.cc]: OK, I have changed it.

**Comment by caojy1998** [comment]
@dgl-bot

File: src/array/cpu/spmat_op_impl_csr.cc [hunk 6]
```diff
@@ -492,10 +492,17 @@ template CSRMatrix CSRSliceRows<kDGLCPU, int64_t>(CSRMatrix, NDArray);
...
+  int64_t new_nrows_temp = rows->shape[0];
+  int64_t new_ncols_temp = cols->shape[0];
+  if (!relabel_nodes) {
+    new_nrows_temp = csr.num_rows;
+    new_ncols_temp = csr.num_cols;
+  }
   IdHashMap<IdType> hashmap(cols);
-  const int64_t new_nrows = rows->shape[0];
-  const int64_t new_ncols = cols->shape[0];
+  const int64_t new_nrows = new_nrows_temp;
```
Comment (Rhett-Ying) [inline src/array/cpu/spmat_op_impl_csr.cc:504]: `num_nrows_temp` is used here for assignment only? If yes, please assign `new_nrows` with lambda or triple expression to remove the `xxx_temp` varaibles

Comment (caojy1998) [reply src/array/cpu/spmat_op_impl_csr.cc]: OK, I changed to triple expression.

File: src/array/cpu/spmat_op_impl_csr.cc [hunk 7]
```diff
@@ -508,16 +509,21 @@ CSRMatrix CSRSliceMatrix(
   std::vector<IdType> sub_data;
   sub_indptr.resize(new_nrows + 1, 0);
   const IdType kInvalidId = new_ncols + 1;
-  for (int64_t i = 0; i < new_nrows; ++i) {
+  for (int64_t i = 0; i < rows->shape[0]; ++i) {
```
Comment (Rhett-Ying) [inline src/array/cpu/spmat_op_impl_csr.cc:512]: As `num_nrows` could be as large as original one if `relabel_nodes=false`, we'd better accelerate slicing via re-design the algorithm or run in parallel with `parallel_for`.

File: tests/cpp/test_spmat_csr.cc [hunk 8]
```diff
@@ -613,6 +613,73 @@ void _TestCSRSliceMatrix1(DGLContext ctx) {
     ASSERT_TRUE(ArrayEQ<IDX>(x.indices, ti));
     ASSERT_TRUE(ArrayEQ<IDX>(x.data, td));
   }
+  if (ctx.device_id == 1) {
```
Comment (Rhett-Ying) [inline tests/cpp/test_spmat_csr.cc:616]: `if (ctx == CPU) return;`?

Comment (caojy1998) [reply tests/cpp/test_spmat_csr.cc]: OK, I have changed it.

**Review by Rhett-Ying — APPROVED** [review]
LGTM

**Comment by caojy1998** [comment]
Now we try to use python API for relabel_nodes=True instread of writing kernel function. The latter is much more complicated.
